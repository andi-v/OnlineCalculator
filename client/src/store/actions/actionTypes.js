// calculator
export const SET_FIRST_NUMBER = 'SET_FIRST_NUMBER'
export const SET_SECOND_NUMBER = 'SET_SECOND_NUMBER'
export const SET_OPERATION = 'SET_OPERATION'
export const CHANGE_CALCULATION_TYPE = 'CHECK_CALCULATION_TYPE'
export const SET_RESULT = 'SET_RESULT'

// history
export const FETCH_HISTORY_START = 'FETCH_HISTORY_START'
export const FETCH_HISTORY_SUCCESS = 'FETCH_HISTORY_SUCCESS'
export const DELETE_HISTORY = 'DELETE_HISTORY'
