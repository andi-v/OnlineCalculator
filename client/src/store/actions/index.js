export {
    setFirstNumber,
    setSecondNumber,
    setOperation,
    changeCalculationType,
    getResult,
    setResult
} from './calculator'
export {
    fetchHistory,
    deleteHistory,
    deleteHistorySuccess
} from './history'