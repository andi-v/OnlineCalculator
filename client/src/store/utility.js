export const updateObject = (oldObject, newProperties) => {
    return {
        ...oldObject,
        ...newProperties
    }
}

export const calculate = (firstNumber, secondNumber, operation) => {
    let result

    switch (operation) {
        case "Addition":
            result = (firstNumber) + (secondNumber);
            break

        case "Subtraction":
            result = (firstNumber) - (secondNumber);
            break

        case "Multiplication":
            result = (firstNumber) * (secondNumber);
            break

        case "Division":
            result = (firstNumber) / (secondNumber);
            break

        case "Power":
            result = Math.pow(firstNumber, secondNumber);
            break
        
        default: break
    }

    return result
}