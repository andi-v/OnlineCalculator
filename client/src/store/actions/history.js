import * as actionTypes from './actionTypes'
import axios from '../../axiosInstance'

export const fetchHistorySuccess = (history) => ({
    type: actionTypes.FETCH_HISTORY_SUCCESS,
    history
})

export const deleteHistorySuccess = () => ({
    type: actionTypes.DELETE_HISTORY_SUCCESS
})

export const deleteHistory = () => {
    return async (dispatch) => {
        try {
            await axios.get('/server_calculator.js/delHistory')

            return dispatch(deleteHistorySuccess())
        }
        catch (err) {
            console.log("error received = ", err)
            return err
        }
    }
}

export const fetchHistory = () => {
    return async (dispatch) => {
        try {
            const response = await axios.get('/server_calculator.js/getHistory')

            dispatch(fetchHistorySuccess(response.data))
        }
        catch (err) {
            console.log("error received = ", err)
            return err
        }
    }
}