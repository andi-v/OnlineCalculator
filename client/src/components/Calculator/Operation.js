import { connect } from 'react-redux'
import { setFirstNumber, setSecondNumber, setOperation } from '../../store/actions'
import Input from './Input'

const Operation = (props) => {
    const {
        firstNumber, secondNumber, operation,
        setFirstNumber, setSecondNumber, setOperation,
    } = props

    return (
        <div>
            <Input
                label="First number:"
                type="text"
                name="firstNumber"
                value={firstNumber}
                onChange={setFirstNumber}
            />
            <Input
                label="Second number:"
                type="text"
                name="secondNumber"
                value={secondNumber}
                onChange={setSecondNumber}
            />

            <strong>Operation :</strong>
            <select
                id="dropdown1"
                onChange={(e) => setOperation(e.target.value)}
                value={operation}
            >
                <option value="Addition">Add (+)</option>
                <option value="Subtraction">Subtract (-)</option>
                <option value="Multiplication">Multiply (*)</option>
                <option value="Division">Divide (/)</option>
                <option value="Power">x to the y power (x^y)</option>
            </select>
        </div>
    )
}

const mapStateToProps = (state) => ({
    firstNumber: state.calc.firstNumber,
    secondNumber: state.calc.secondNumber,
    operation: state.calc.operation
})

const mapDispatchToProps = ({
    setFirstNumber,
    setSecondNumber,
    setOperation
})

export default connect(mapStateToProps, mapDispatchToProps)(Operation)