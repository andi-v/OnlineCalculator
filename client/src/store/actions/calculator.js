import * as actionTypes from './actionTypes'
import { calculate } from '../utility'

export const setFirstNumber = (firstNumber) => ({
    type: actionTypes.SET_FIRST_NUMBER,
    number: firstNumber
})

export const setSecondNumber = (secondNumber) => ({
    type: actionTypes.SET_SECOND_NUMBER,
    number: secondNumber
})

export const setOperation = (operation) => ({
    type: actionTypes.SET_OPERATION,
    operation: operation
})

export const changeCalculationType = () => ({ type: actionTypes.CHANGE_CALCULATION_TYPE })

const calculateResult = async (firstNumber, secondNumber, operation, serverCalculation) => {
    if (serverCalculation) {
        try {
            const response = await window.axios.get('/server_calculator.js', {
                params: {
                    firstNumber,
                    secondNumber,
                    operation
                }
            })

            return response.data
        }
        catch (err) {
            return err
        }
    }
    else {
        const result = calculate(+firstNumber, +secondNumber, operation)

        await window.axios.get('/server_calculator.js', {
            params: { firstNumber, secondNumber, operation, result }
        })

        return result
    }
}

export const setResult = (result) => {
    return {
        type: actionTypes.SET_RESULT,
        result
    }
}

export const getResult = () => {
    return async (dispatch, getState) => {
        const { firstNumber, secondNumber, operation, serverCalculation } = getState().calc
        const result = await calculateResult(firstNumber, secondNumber, operation, serverCalculation)

        console.log("response = ", Number(result))
        dispatch(setResult(result))
    }
}