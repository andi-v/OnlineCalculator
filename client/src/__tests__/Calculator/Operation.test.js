import React from 'react'
import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import { setFirstNumber, setSecondNumber, setOperation } from '../../store/actions/'
import Operation from '../../components/Calculator/Operation'
import toJSON from 'enzyme-to-json'

const middlewares = []
const mockStore = configureStore(middlewares)

const initialState = {
    calc: {
        firstNumber: '0',
        secondNumber: '0',
        operation: "Addition"
    }
}
const store = mockStore(initialState)

const wrapper = mount(
    <Provider store={store}>
        <Operation />
    </Provider>
)

describe("The Operation component rendering", () => {
    it("renders with given state from mocked store", () => {
        expect(toJSON(wrapper)).toMatchSnapshot()
    })

    it("displays the right label for first number input", () => {
        const label = wrapper.find("Input").at(0).prop("label")
        expect(label).toBe("First number:")
    })

    it("displays the value from state for the first number", () => {
        const firstNumber = wrapper.find("Input").at(0).prop("value")
        expect(firstNumber).toBe(initialState.calc.firstNumber)
    })

    it("displays the right label for second number input", () => {
        const label = wrapper.find("Input").at(1).prop("label")
        expect(label).toBe("Second number:")
    })

    it("displays the value from state for the second number", () => {
        const secondNumber = wrapper.find("Input").at(1).prop("value")
        expect(secondNumber).toBe(initialState.calc.secondNumber)
    })

    it("displays the right label for operation dropdown", () => {
        const label = wrapper.text()
        expect(label).toEqual(expect.stringContaining("Operation :"))
    })

    it("displays the value from state as the default operation value", () => {
        const selectValue = wrapper.find("select").prop("value")
        expect(selectValue).toBe(initialState.calc.operation)
    })
})

describe("The Operation component dispatching the right actions", () => {
    const expectedActions = [
        setFirstNumber("23"),
        setSecondNumber("71"),
        setOperation("Addition")
    ]

    it("dispatches the right action when changing first number", () => {
        wrapper.find("Operation").props().setFirstNumber("23")
        expect(store.getActions()).toContainEqual(expectedActions[0])
    })

    it("dispatches the right action when changing second number", () => {
        wrapper.find("Operation").props().setSecondNumber("71")
        expect(store.getActions()).toContainEqual(expectedActions[1])
    })

    it("dispatches the right action when changing operation", () => {
        wrapper.find("Operation").props().setOperation("Addition")
        expect(store.getActions()).toContainEqual(expectedActions[2])
    })
})
