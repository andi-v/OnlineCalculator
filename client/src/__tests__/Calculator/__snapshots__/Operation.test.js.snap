// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Operation component rendering renders with given state from mocked store 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Operation)>
    <Operation
      firstNumber="0"
      operation="Addition"
      secondNumber="0"
      setFirstNumber={[Function]}
      setOperation={[Function]}
      setSecondNumber={[Function]}
    >
      <div>
        <Input
          label="First number:"
          name="firstNumber"
          onChange={[Function]}
          type="text"
          value="0"
        >
          <label>
            <label
              htmlFor="firstNumber"
            >
              First number:
            </label>
            <input
              id="firstNumber"
              name="firstNumber"
              onChange={[Function]}
              type="text"
              value="0"
            />
          </label>
        </Input>
        <Input
          label="Second number:"
          name="secondNumber"
          onChange={[Function]}
          type="text"
          value="0"
        >
          <label>
            <label
              htmlFor="secondNumber"
            >
              Second number:
            </label>
            <input
              id="secondNumber"
              name="secondNumber"
              onChange={[Function]}
              type="text"
              value="0"
            />
          </label>
        </Input>
        Operation :
        <select
          id="dropdown1"
          onChange={[Function]}
          value="Addition"
        >
          <option
            value="Addition"
          >
            Add (+)
          </option>
          <option
            value="Subtraction"
          >
            Subtract (-)
          </option>
          <option
            value="Multiplication"
          >
            Multiply (*)
          </option>
          <option
            value="Division"
          >
            Divide (/)
          </option>
          <option
            value="Power"
          >
            x to the y power (x^y)
          </option>
        </select>
      </div>
    </Operation>
  </Connect(Operation)>
</Provider>
`;
